doc/man/man3/Update.3
.TH "Update" 3 "Fri Sep 1 2017" "Version 1.0" "COM-Express I2C Stack" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Update \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <update\&.hpp>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBUpdate\fP (\fBI2C_base\fP *i2c)"
.br
.RI "\fIClass constructor\&. \fP"
.ti -1c
.RI "\fB~Update\fP (void)"
.br
.RI "\fIClass destructor\&. \fP"
.ti -1c
.RI "void \fBsaveActive\fP (void)"
.br
.RI "\fISave active registers to local cache and then push the database to an external server\&. \fP"
.ti -1c
.RI "void \fBsaveStatic\fP (void)"
.br
.RI "\fISave static registers to local cache and then push the database to an external server\&. \fP"
.ti -1c
.RI "void \fBpurgeDatabase\fP (void)"
.br
.RI "\fIPurge database of any entries over an hour old\&. \fP"
.ti -1c
.RI "void \fBgetConfig\fP (void)"
.br
.RI "\fICheck if config files have changed and update registers if true\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 19 of file update\&.hpp\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Update::Update (\fBI2C_base\fP * i2c)"

.PP
Class constructor\&. 
.PP
Definition at line 13 of file update\&.cpp\&.
.SS "Update::~Update (void)"

.PP
Class destructor\&. 
.PP
Definition at line 22 of file update\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void Update::getConfig (void)"

.PP
Check if config files have changed and update registers if true\&. 
.PP
Definition at line 156 of file update\&.cpp\&.
.SS "void Update::purgeDatabase (void)"

.PP
Purge database of any entries over an hour old\&. 
.PP
Definition at line 116 of file update\&.cpp\&.
.SS "void Update::saveActive (void)"

.PP
Save active registers to local cache and then push the database to an external server\&. 
.PP
Definition at line 29 of file update\&.cpp\&.
.SS "void Update::saveStatic (void)"

.PP
Save static registers to local cache and then push the database to an external server\&. 
.PP
Definition at line 60 of file update\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for COM-Express I2C Stack from the source code\&.


src/update.cpp
/**
  @file update.cpp
  @brief Defines functions for the Update class.
  @author David Monk - Imperial College London
  @version 1.0
*/

#include "update.hpp"

/**
   @brief Class constructor.
*/
Update::Update(I2C_base* i2c) {
   board = new ATCABoard(i2c);
   std::ifstream config_file("/root/I2C/bin/config.json");
   config_file >> config;
}

/**
   @brief Class destructor.
*/
Update::~Update(void) {
   delete board;
}

/**
   @brief Save active registers to local cache and then push the database to an external server.
*/
void Update::saveActive(void) {
   json board_dict;
   board_dict["_id"] = getTime();
   std::ifstream data_file("/root/I2C/bin/active.json");
   json active;
   data_file >> active;
   for (auto bus : board->getBuses()) {
      board_dict[bus] = {};
      board->setBus(bus);
      for (auto device : board->getDevices()) {
         board_dict[bus][device] = {};
         board->setDevice(device);
         for (auto property : board->getProperties()) {
            if (active[bus][device][property]) {
               board_dict[bus][device][property] = board->read(property);
            }
         }
      }
   }
   Client* client = new Client();
   client->setDatabase("data");
   client->uploadDocument(board_dict);
   if(client->checkOnline(config["target"]["url"].get<std::string>())) {
      client->pushDatabase();
   }
   delete client;
}

/**
   @brief Save static registers to local cache and then push the database to an external server.
*/
void Update::saveStatic(void) {
   Client* client = new Client();
   client->setDatabase("data");
   bool flag = false;
   json doc = json::parse(client->getDocument("static"));
   json metadata, board_dict, active;
   board_dict["_id"] = doc["_id"];
   board_dict["_rev"] = doc["_rev"];
   std::ifstream data_file("/root/I2C/bin/active.json");
   data_file >> active;
   for (auto bus : board->getBuses()) {
      board_dict[bus] = {};
      board->setBus(bus);
      for (auto device : board->getDevices()) {
         board_dict[bus][device] = {};
         board->setDevice(device);
         for (auto property : board->getProperties()) {
            if (!active[bus][device][property]) {
               std::string reg = board->read(property);
               board_dict[bus][device][property] = reg;
               if (reg != doc[bus][device][property]) {
                  flag = true;
                  metadata.push_back({bus, device, property});
               }
            }
         }
      }
   }
   if (flag) {
      sendFlag(board_dict, metadata, client);
   }
   client->uploadDocument(board_dict);
   if(client->checkOnline(config["target"]["url"].get<std::string>())) {
      client->pushDatabase();
   }
   delete client;
}

/**
   @brief Send flag to external server if static register is changed.
   @param data - register data dump at time of change.
   @param metadata - JSON of metadata for flag.
   @param client - CouchDB Client class to send data to server.
*/
void Update::sendFlag(json data, json metadata, Client* client) {
   json flag_dict = {
      {"_id",config["target"]["dbname"].get<std::string>() + "-" + getTime()},
      {"properties", data},
      {"metadata", metadata}
   };
   client->uploadDocument(config["target"]["url"].get<std::string>()+"/flags",flag_dict);
}

/**
   @brief Purge database of any entries over an hour old.
*/
void Update::purgeDatabase(void) {
   Client* client = new Client();
   client->setDatabase("data");
   if(client->checkOnline(config["target"]["url"].get<std::string>())) {
      int time_int = atoi(getTime().c_str());
      for (auto i : client->getDocumentIDs()) {
         int ID_int = atoi(i.first.c_str());
         if ((time_int - ID_int) > 10000) {
            client->deleteDocument(i.first, i.second);
         }
      }
      client->compactDatabase();
   }
   delete client;
}

/**
   @brief Write config database to registers on card.
*/
void Update::writeConfig(void) {
   Client* client = new Client();
   client->setDatabase("config");
   json doc = json::parse(client->getDocument(config["target"]["dbname"]));
   delete client;
   for (json::iterator bus = doc.begin(); bus != doc.end(); ++bus) {
      if (bus.key().at(0) != '_') {
         board->setBus(bus.key());
         for (json::iterator device = bus.value().begin(); device != bus.value().end(); ++device) {
            board->setDevice(device.key());
            for (json::iterator property = device.value().begin(); property != device.value().end(); ++property) {
               board->write(property.key(), property.value());
            }
         }
      }
   }
}

/**
   @brief Check if config files have changed and update registers if true.
*/
void Update::getConfig(void) {
   std::string filename = config["config_db"]["path"].get<std::string>();
   int rev = (int)boost::filesystem::last_write_time(filename.c_str());
   if (rev != config["config_db"]["rev"].get<int>()) {
      std::cout << "Changing registers..." << '\n';
      writeConfig();
      config["config_db"]["rev"] = rev;
      std::ofstream config_file("/root/I2C/bin/config.json");
      config_file << std::setw(4) << config << std::endl;
   }
}

std::string Update::getTime(void) {
   auto t = std::time(nullptr);
   auto tm = *std::localtime(&t);
   std::ostringstream oss;
   oss << std::put_time(&tm, "%Y%m%d%H%M%S");
   return oss.str();
}
